src/csharp/Grpc.HealthCheck/HealthServiceImpl.cs

```diff
@@ -51,7 +54,13 @@ public void SetStatus(string service, HealthCheckResponse.Types.ServingStatus st
         {
             lock (myLock)
             {
+                HealthCheckResponse.Types.ServingStatus previousStatus = GetServiceStatus(service);
                 statusMap[service] = status;
+
+                if (status != previousStatus)
+                {
+                    NotifyStatus(service, status);
```

Comment:
I don't think the NotifyStatus (which sends out responses to watchers) should be called from under the lock.
Also below.